addons/mrp/models/mrp_production.py

```diff
@@ -1205,6 +1217,37 @@ def action_confirm(self):
         self.filtered(lambda mo: mo.state == 'draft').state = 'confirmed'
         return True
 
+    def _link_workorders_and_moves(self):
+        self.ensure_one()
+        if not self.workorder_ids:
+            return
+        workorder_per_operation = {workorder.operation_id: workorder for workorder in self.workorder_ids}
+        self.allow_workorder_dependencies = self.bom_id.allow_operation_dependencies
+        if self.allow_workorder_dependencies:
+            for workorder in self.workorder_ids:
+                workorder.blocked_by_workorder_ids = [Command.link(workorder_per_operation[operation_id].id) for operation_id in workorder.operation_id.blocked_by_operation_ids]
+        else:
+            previous_workorder = self.env['mrp.production']
+            for workorder in self.workorder_ids:
+                if previous_workorder and previous_workorder.operation_id.bom_id == workorder.operation_id.bom_id:
+                    workorder.blocked_by_workorder_ids = [Command.link(previous_workorder.id)]
+                previous_workorder = workorder
+        boms = set(workorder.operation_id.bom_id for workorder in self.workorder_ids)
+        last_workorder_per_bom = {bom: self.workorder_ids.filtered(lambda wo: wo.operation_id.bom_id == bom and not wo.needed_by_workorder_ids)[-1] for bom in boms}
```

Comment:
Maybe we can avoid looping through all of the workorders a second time by storing the value during the first loop with:

```
if self.allow_workorder_dependencies:
    for workorder in self.workorder_ids:
        workorder.blocked_by_workorder_ids = [Command.link(workorder_per_operation[operation_id].id) for operation_id in workorder.operation_id.blocked_by_operation_ids]
else:
    previous_workorder = self.env['mrp.production']
    for workorder in self.workorder_ids:
        if previous_workorder and previous_workorder.operation_id.bom_id == workorder.operation_id.bom_id:
            workorder.blocked_by_workorder_ids = [Command.link(previous_workorder.id)]
        previous_workorder = workorder
```

by adding before the for loops:
`last_workorder_per_bom = defaultdict(lambda: self.env['mrp.workorder']` (or `last_workorder_per_bom = defaultdict(int)` since it looks like only the id is being used in the end) 

and

```
if not workorder.blocked_by_workorder_ids:
   last_workorder_per_bom[workorder.operation_id.bom_id] = workorder
```
within each for loop